File: cli_kanban_go/internal/tui/update.go
package tui

import (
	"strings"
	"time"

	"github.com/charmbracelet/bubbles/viewport"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/happytaoer/cli_kanban/internal/model"
)

// Update handles messages and updates the model
func (m Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	var cmd tea.Cmd

	switch msg := msg.(type) {
	case tea.WindowSizeMsg:
		m.width = msg.Width
		m.height = msg.Height

		// Calculate viewport height (total height - header - footer)
		headerHeight := 2 // title+stats line + spacing
		footerHeight := 3 // footer + spacing
		vpHeight := msg.Height - headerHeight - footerHeight
		if vpHeight < 1 {
			vpHeight = 1
		}

		if !m.ready {
			m.viewport = viewport.New(msg.Width, vpHeight)
			m.viewport.YPosition = headerHeight
			m.ready = true
		} else {
			m.viewport.Width = msg.Width
			m.viewport.Height = vpHeight
		}
		return m, nil

	case clockTickMsg:
		m.currentTime = time.Time(msg)
		return m, clockTickCmd()

	case tasksLoadedMsg:
		m.organizeTasks(msg.tasks)
		return m, nil

	case taskCreatedMsg:
		return m, m.loadTasks()

	case taskUpdatedMsg:
		return m, m.loadTasks()

	case taskDeletedMsg:
		return m, m.loadTasks()

	case descriptionUpdatedMsg:
		return m, m.loadTasks()

	case tagsUpdatedMsg:
		return m, m.loadTasks()

	case errMsg:
		m.err = msg.err
		return m, nil

	case tea.KeyMsg:
		return m.handleKeyPress(msg)
	}

	// Handle text input updates
	if m.viewMode == ViewModeAddTask || m.viewMode == ViewModeEditTask || m.viewMode == ViewModeEditTags {
		m.textInput, cmd = m.textInput.Update(msg)
		return m, cmd
	}

	// Handle textarea updates
	if m.viewMode == ViewModeEditDescription {
		m.textArea, cmd = m.textArea.Update(msg)
		return m, cmd
	}

	return m, nil
}

// handleKeyPress handles keyboard input
func (m Model) handleKeyPress(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	// Global keys
	switch msg.String() {
	case "ctrl+c", "q":
		if m.viewMode == ViewModeBoard {
			return m, tea.Quit
		}
	case "esc":
		if m.viewMode != ViewModeBoard {
			m.viewMode = ViewModeBoard
			m.textInput.SetValue("")
			return m, nil
		}
		return m, tea.Quit
	}

	// Mode-specific keys
	switch m.viewMode {
	case ViewModeBoard:
		return m.handleBoardKeys(msg)
	case ViewModeAddTask:
		return m.handleAddTaskKeys(msg)
	case ViewModeEditTask:
		return m.handleEditTaskKeys(msg)
	case ViewModeEditDescription:
		return m.handleEditDescriptionKeys(msg)
	case ViewModeEditTags:
		return m.handleEditTagsKeys(msg)
	case ViewModeConfirmDelete:
		return m.handleConfirmDeleteKeys(msg)
	case ViewModeHelp:
		return m.handleHelpKeys(msg)
	}

	return m, nil
}

// handleBoardKeys handles keyboard input in board view mode
func (m Model) handleBoardKeys(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	switch msg.String() {
	case "left", "h":
		if m.currentColumn > 0 {
			m.currentColumn--
			m.currentTask = 0
		}
		return m, nil

	case "right", "l":
		if m.currentColumn < len(m.columns)-1 {
			m.currentColumn++
			m.currentTask = 0
		}
		return m, nil

	case "up", "k":
		if m.currentTask > 0 {
			m.currentTask--
			m.ensureTaskVisible()
		}
		return m, nil

	case "down", "j":
		col := m.columns[m.currentColumn]
		if m.currentTask < len(col.Tasks)-1 {
			m.currentTask++
			m.ensureTaskVisible()
		}
		return m, nil

	case "a":
		m.viewMode = ViewModeAddTask
		m.textInput.SetValue("")
		m.textInput.Focus()
		return m, nil

	case "e", "enter":
		task := m.getCurrentTask()
		if task != nil {
			m.viewMode = ViewModeEditTask
			m.textInput.SetValue(task.Title)
			m.textInput.Focus()
		}
		return m, nil

	case "d", "delete":
		task := m.getCurrentTask()
		if task != nil {
			m.pendingDeleteID = task.ID
			m.viewMode = ViewModeConfirmDelete
		}
		return m, nil

	case "m":
		task := m.getCurrentTask()
		if task != nil {
			nextColumn := (m.currentColumn + 1) % len(m.columns)
			m.currentColumn = nextColumn
			m.followTaskID = task.ID
			return m, m.moveTask(task, nextColumn)
		}
		return m, nil

	case "i":
		task := m.getCurrentTask()
		if task != nil {
			m.viewMode = ViewModeEditDescription
			m.textArea.SetValue(task.Description)
			m.textArea.Focus()
		}
		return m, nil

	case "t":
		task := m.getCurrentTask()
		if task != nil {
			m.viewMode = ViewModeEditTags
			m.textInput.SetValue(strings.Join(task.Tags, ", "))
			m.textInput.Focus()
		}
		return m, nil

	case "?":
		m.viewMode = ViewModeHelp
		return m, nil
	}

	return m, nil
}

// handleAddTaskKeys handles keyboard input in add task mode
func (m Model) handleAddTaskKeys(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	switch msg.String() {
	case "enter":
		title := m.textInput.Value()
		if title != "" {
			status := m.columns[m.currentColumn].Status
			m.viewMode = ViewModeBoard
			m.textInput.SetValue("")
			return m, m.createTask(title, status)
		}
		return m, nil

	case "esc":
		m.viewMode = ViewModeBoard
		m.textInput.SetValue("")
		return m, nil
	}

	var cmd tea.Cmd
	m.textInput, cmd = m.textInput.Update(msg)
	return m, cmd
}

// handleEditTaskKeys handles keyboard input in edit task mode
func (m Model) handleEditTaskKeys(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	switch msg.String() {
	case "enter":
		title := m.textInput.Value()
		task := m.getCurrentTask()
		if title != "" && task != nil {
			m.viewMode = ViewModeBoard
			m.textInput.SetValue("")
			return m, m.updateTask(task.ID, title, task.Status)
		}
		return m, nil

	case "esc":
		m.viewMode = ViewModeBoard
		m.textInput.SetValue("")
		return m, nil
	}

	var cmd tea.Cmd
	m.textInput, cmd = m.textInput.Update(msg)
	return m, cmd
}

// handleEditDescriptionKeys handles keyboard input in edit description mode
func (m Model) handleEditDescriptionKeys(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	switch msg.String() {
	case "ctrl+s":
		description := m.textArea.Value()
		task := m.getCurrentTask()
		if task != nil {
			m.viewMode = ViewModeBoard
			m.textArea.SetValue("")
			return m, m.updateDescription(task.ID, description)
		}
		return m, nil

	case "esc":
		m.viewMode = ViewModeBoard
		m.textArea.SetValue("")
		return m, nil
	}

	var cmd tea.Cmd
	m.textArea, cmd = m.textArea.Update(msg)
	return m, cmd
}

// handleEditTagsKeys handles keyboard input in edit tags mode
func (m Model) handleEditTagsKeys(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	switch msg.String() {
	case "enter":
		tagsStr := m.textInput.Value()
		task := m.getCurrentTask()
		if task != nil {
			tags := parseTagsInput(tagsStr)
			m.viewMode = ViewModeBoard
			m.textInput.SetValue("")
			return m, m.updateTags(task.ID, tags)
		}
		return m, nil

	case "esc":
		m.viewMode = ViewModeBoard
		m.textInput.SetValue("")
		return m, nil
	}

	var cmd tea.Cmd
	m.textInput, cmd = m.textInput.Update(msg)
	return m, cmd
}

// parseTagsInput parses comma-separated tags input
func parseTagsInput(input string) []string {
	parts := strings.Split(input, ",")
	var tags []string
	for _, p := range parts {
		t := strings.TrimSpace(strings.ToLower(p))
		if t != "" {
			tags = append(tags, t)
		}
	}
	return tags
}

// handleConfirmDeleteKeys handles keyboard input in delete confirmation mode
func (m Model) handleConfirmDeleteKeys(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	switch msg.String() {
	case "y", "Y":
		id := m.pendingDeleteID
		m.pendingDeleteID = 0
		m.viewMode = ViewModeBoard
		return m, m.deleteTask(id)

	case "n", "N", "esc":
		m.pendingDeleteID = 0
		m.viewMode = ViewModeBoard
		return m, nil
	}

	return m, nil
}

// handleHelpKeys handles keyboard input in help mode
func (m Model) handleHelpKeys(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	m.viewMode = ViewModeBoard
	return m, nil
}

// createTask creates a new task
func (m Model) createTask(title string, status model.TaskStatus) tea.Cmd {
	return func() tea.Msg {
		task, err := m.db.CreateTask(title, status)
		if err != nil {
			return errMsg{err}
		}
		return taskCreatedMsg{task}
	}
}

// updateTask updates a task
func (m Model) updateTask(id int64, title string, status model.TaskStatus) tea.Cmd {
	return func() tea.Msg {
		err := m.db.UpdateTask(id, title, status)
		if err != nil {
			return errMsg{err}
		}
		return taskUpdatedMsg{}
	}
}

// deleteTask deletes a task
func (m Model) deleteTask(id int64) tea.Cmd {
	return func() tea.Msg {
		err := m.db.DeleteTask(id)
		if err != nil {
			return errMsg{err}
		}
		return taskDeletedMsg{}
	}
}

// updateDescription updates a task's description
func (m Model) updateDescription(id int64, description string) tea.Cmd {
	return func() tea.Msg {
		err := m.db.UpdateTaskDescription(id, description)
		if err != nil {
			return errMsg{err}
		}
		return descriptionUpdatedMsg{}
	}
}

// updateTags updates a task's tags
func (m Model) updateTags(id int64, tags []string) tea.Cmd {
	return func() tea.Msg {
		err := m.db.UpdateTaskTags(id, tags)
		if err != nil {
			return errMsg{err}
		}
		return tagsUpdatedMsg{}
	}
}

// moveTask moves a task to the target column
func (m Model) moveTask(task *model.Task, targetColumn int) tea.Cmd {
	newStatus := m.columns[targetColumn].Status

	return func() tea.Msg {
		err := m.db.UpdateTaskStatus(task.ID, newStatus)
		if err != nil {
			return errMsg{err}
		}
		return taskUpdatedMsg{}
	}
}

File: cli_kanban_python/cli_kanban/tui/update.py
"""TUI update/input handler for Kanban board."""

from cli_kanban.model import TaskStatus
from .model import TuiModel, ViewMode


def handle_key_press(model: TuiModel, key: str) -> bool:
    """Handle keyboard input.
    
    Args:
        model: TUI model to update.
        key: Key pressed (can be char, 'left', 'right', 'up', 'down', 'enter', 'esc', etc).
        
    Returns:
        True if should quit, False otherwise.
    """
    # Global keys
    if key == "q" or key == "ctrl+c":
        if model.view_mode == ViewMode.BOARD:
            return True
    
    if key == "esc":
        if model.view_mode != ViewMode.BOARD:
            model.view_mode = ViewMode.BOARD
            model.text_input = ""
        else:
            return True

    # Mode-specific handling
    if model.view_mode == ViewMode.BOARD:
        return handle_board_keys(model, key)
    elif model.view_mode == ViewMode.ADD_TASK:
        return handle_add_task_keys(model, key)
    elif model.view_mode == ViewMode.EDIT_TASK:
        return handle_edit_task_keys(model, key)
    elif model.view_mode == ViewMode.EDIT_DESCRIPTION:
        return handle_edit_description_keys(model, key)
    elif model.view_mode == ViewMode.EDIT_TAGS:
        return handle_edit_tags_keys(model, key)
    elif model.view_mode == ViewMode.CONFIRM_DELETE:
        return handle_confirm_delete_keys(model, key)
    elif model.view_mode == ViewMode.HELP:
        return handle_help_keys(model, key)

    return False


def handle_board_keys(model: TuiModel, key: str) -> bool:
    """Handle keys in board view mode."""
    if key == "left" or key == "h":
        model.navigate_left()
    elif key == "right" or key == "l":
        model.navigate_right()
    elif key == "up" or key == "k":
        model.navigate_up()
    elif key == "down" or key == "j":
        model.navigate_down()
    elif key == "a":
        model.view_mode = ViewMode.ADD_TASK
        model.text_input = ""
    elif key == "e" or key == "enter":
        task = model.get_current_task()
        if task:
            model.view_mode = ViewMode.EDIT_TASK
            model.text_input = task.title
    elif key == "d" or key == "delete":
        task = model.get_current_task()
        if task:
            model.pending_delete_id = task.id
            model.view_mode = ViewMode.CONFIRM_DELETE
    elif key == "m":
        task = model.get_current_task()
        if task:
            next_column = (model.current_column + 1) % len(model.columns)
            model.current_column = next_column
            model.follow_task_id = task.id
            model.move_task(task, next_column)
    elif key == "i":
        task = model.get_current_task()
        if task:
            model.view_mode = ViewMode.EDIT_DESCRIPTION
            model.text_input = task.description
    elif key == "t":
        task = model.get_current_task()
        if task:
            model.view_mode = ViewMode.EDIT_TAGS
            model.text_input = ", ".join(task.tags)
    elif key == "?":
        model.view_mode = ViewMode.HELP

    return False


def handle_add_task_keys(model: TuiModel, key: str) -> bool:
    """Handle keys in add task mode."""
    if key == "enter":
        if model.text_input.strip():
            status = model.columns[model.current_column].status
            model.create_task(model.text_input.strip(), status)
            model.view_mode = ViewMode.BOARD
            model.text_input = ""
    elif key == "esc":
        model.view_mode = ViewMode.BOARD
        model.text_input = ""
    elif key == "backspace":
        model.text_input = model.text_input[:-1]
    elif len(key) == 1 and ord(key) >= 32:  # Printable character
        model.text_input += key

    return False


def handle_edit_task_keys(model: TuiModel, key: str) -> bool:
    """Handle keys in edit task mode."""
    if key == "enter":
        if model.text_input.strip():
            task = model.get_current_task()
            if task:
                model.update_task_title(task.id, model.text_input.strip())
                model.view_mode = ViewMode.BOARD
                model.text_input = ""
    elif key == "esc":
        model.view_mode = ViewMode.BOARD
        model.text_input = ""
    elif key == "backspace":
        model.text_input = model.text_input[:-1]
    elif len(key) == 1 and ord(key) >= 32:  # Printable character
        model.text_input += key

    return False


def handle_edit_description_keys(model: TuiModel, key: str) -> bool:
    """Handle keys in edit description mode."""
    if key == "ctrl+s":
        task = model.get_current_task()
        if task:
            model.update_task_description(task.id, model.text_input)
            model.view_mode = ViewMode.BOARD
            model.text_input = ""
    elif key == "esc":
        model.view_mode = ViewMode.BOARD
        model.text_input = ""
    elif key == "backspace":
        model.text_input = model.text_input[:-1]
    elif key == "enter":
        model.text_input += "\n"
    elif len(key) == 1 and ord(key) >= 32:  # Printable character
        model.text_input += key

    return False


def handle_edit_tags_keys(model: TuiModel, key: str) -> bool:
    """Handle keys in edit tags mode."""
    if key == "enter":
        task = model.get_current_task()
        if task:
            tags = parse_tags_input(model.text_input)
            model.update_task_tags(task.id, tags)
            model.view_mode = ViewMode.BOARD
            model.text_input = ""
    elif key == "esc":
        model.view_mode = ViewMode.BOARD
        model.text_input = ""
    elif key == "backspace":
        model.text_input = model.text_input[:-1]
    elif len(key) == 1 and ord(key) >= 32:  # Printable character
        model.text_input += key

    return False


def handle_confirm_delete_keys(model: TuiModel, key: str) -> bool:
    """Handle keys in delete confirmation mode."""
    if key == "y":
        model.delete_task(model.pending_delete_id)
        model.pending_delete_id = 0
        model.view_mode = ViewMode.BOARD
    elif key == "n" or key == "esc":
        model.pending_delete_id = 0
        model.view_mode = ViewMode.BOARD

    return False


def handle_help_keys(model: TuiModel, key: str) -> bool:
    """Handle keys in help mode."""
    model.view_mode = ViewMode.BOARD
    return False


def parse_tags_input(input_str: str) -> list:
    """Parse comma-separated tags input."""
    tags = []
    for tag in input_str.split(","):
        tag = tag.strip().lower()
        if tag:
            tags.append(tag)
    return tags
